src/unsync/mpsc.rs

```diff
@@ -0,0 +1,127 @@
+pub fn channel<T>(buffer: usize) -> (Sender<T>, Receiver<T>) {
+    let shared = Rc::new(RefCell::new(Shared {
+        buffer: VecDeque::with_capacity(buffer),
+        receiver_live: true,
+        blocked_senders: VecDeque::with_capacity(1),
+        blocked_recv: None,
+    }));
+    let sender = Sender { shared: shared.clone() };
+    (sender, Receiver { shared: Some(shared) })
+}
+
+struct Shared<T> {
+    buffer: VecDeque<T>,
+    receiver_live: bool,
+    blocked_senders: VecDeque<Task>,
+    blocked_recv: Option<Task>,
+}
+
+impl<T> Shared<T> {
+    fn send(&mut self, msg: T) -> AsyncSink<T> {
+        if self.buffer.len() == self.buffer.capacity() {
+            self.blocked_senders.push_back(task::park());
+            AsyncSink::NotReady(msg)
+        } else {
+            self.buffer.push_back(msg);
+            if let Some(task) = self.blocked_recv.take() {
+                task.unpark();
+            }
+            AsyncSink::Ready
+        }
+    }
+
+    fn recv(&mut self) -> Async<T> {
+        match self.buffer.pop_front() {
+            None => {
+                self.blocked_recv = Some(task::park());
+                Async::NotReady
+            }
+            Some(msg) => {
+                if let Some(task) = self.blocked_senders.pop_front() {
+                    task.unpark();
+                }
+                Async::Ready(msg)
+            }
+        }
+    }
+}
+
+/// The transmission end of a channel.
+///
+/// This is created by the `channel` function.
+pub struct Sender<T> {
+    shared: Rc<RefCell<Shared<T>>>,    
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Sender<T> { Sender { shared: self.shared.clone() } }
+}
+
+impl<T> Sink for Sender<T> {
+    type SinkItem = T;
+    type SinkError = SendError<T>;
+
+    fn start_send(&mut self, msg: T) -> StartSend<T, SendError<T>> {
+        let mut shared = self.shared.borrow_mut();
+        if !shared.receiver_live {
+            Err(SendError(msg))
+        } else {
+            Ok(shared.send(msg))
+        }
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), SendError<T>> {
+        Ok(Async::Ready(()))
+    }
+}
+
+
+/// The receiving end of a channel which implements the `Stream` trait.
+///
+/// This is created by the `channel` function.
+pub struct Receiver<T> {
+    shared: Option<Rc<RefCell<Shared<T>>>>,
+}
+
+impl<T> Drop for Receiver<T> {
+    fn drop(&mut self) {
+        if let Some(shared) = self.shared.as_ref() {
+            shared.borrow_mut().receiver_live = false;
+        }
+    }
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+    type Error = ();            // TODO: Replace with uninhabited type, as this can never fail
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if let Async::Ready(msg) = self.shared.as_ref().unwrap().borrow_mut().recv() {
+            return Ok(Async::Ready(Some(msg)));
+        }
+
+        match Rc::try_unwrap(self.shared.take().unwrap()) {
```

Comment:
The ownership movement here seems sort of unfortunate, but I think we've got some other strategies:

* Use `get_mut` instead of this to test "unique ownership"
* My preferred strategy, store a weak pointer in the `Receiver`, and a failed upgrade means the channel is terminated.

One case here I think that isn't handled is where once all senders are dropped it should unpark the receiver, and that could be handled naturally with a `Drop` implementation if the receiver stores a weak pointer.